cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,277 @@
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP && iph->protocol != IPPROTO_UDP)
+        return TC_ACT_OK;
+
+    if (!capture_dns && iph->protocol == IPPROTO_UDP) {
+        if (data + sizeof(*eth) + sizeof(*iph) + sizeof(*udph) > data_end)
+            return TC_ACT_OK;
+        udph = data + sizeof(*eth) + sizeof(*iph);
+        if (udph->dest == bpf_htons(UDP_P_DNS))
+            return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = OUTBOUND_CB;
+
+    dbg("outbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For app pod veth pair(in host ns) egress hook
+SEC("tc")
+int app_inbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr  *eth = data;
+    struct iphdr *iph = NULL;
+    struct udphdr *udph;
+    struct ztunnel_info *zi = NULL;
+
+    if (skb->cb[4] == BYPASS_CB)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP && iph->protocol != IPPROTO_UDP)
+        return TC_ACT_OK;
+
+    zi = get_ztunnel_info();
+
+    if (!zi || zi->ifindex == 0) {
+        dbg("inbound_to_ztunnel unable to retrieve ztunnel_info\n");
+        return TC_ACT_OK;
+    }
+
+    if (iph->protocol == IPPROTO_UDP) {
+        if (data + sizeof(*eth) + sizeof(*iph) + sizeof(*udph) > data_end)
+            return TC_ACT_OK;
+        udph = data + sizeof(*eth) + sizeof(*iph);
+        if (udph->source == bpf_htons(UDP_P_DNS))
+            return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = INBOUND_CB;
+
+    dbg("inbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For ztunnel pod veth pair(in host ns) ingress hook
+SEC("tc")
+int ztunnel_host_ingress(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct iphdr *iph = NULL;
+    struct app_info *pi = NULL;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP && iph->protocol != IPPROTO_UDP)
+        return TC_ACT_OK;
+
+    pi = get_app_info_from_ipv4(iph->daddr);
```

Comment:
So we should be able to tell (dynamically) if this is IPV6 vs 4 from `protocol` above, and decide what to use to lookup -> `protocol == ETH_P_IP || protocol == ETH_P_IPV6`

Reply:
To limit the scope of the PR, IPv6 and dual stack is the next step.